pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConnectionController.java

```diff
@@ -0,0 +1,133 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.apache.pulsar.broker.ServiceConfiguration;
+import org.apache.pulsar.common.tls.InetAddressUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public interface ConnectionController {
+
+    /**
+     * Increase the number of connections counter.
+     * @param remoteAddress
+     * @return
+     */
+    boolean increaseConnection(SocketAddress remoteAddress);
+
+    /**
+     * Decrease the number of connections counter.
+     * @param remoteAddress
+     */
+    void decreaseConnection(SocketAddress remoteAddress);
+
+
+    class DefaultConnectionController implements ConnectionController {
+        private static final Logger log = LoggerFactory.getLogger(DefaultConnectionController.class);
+        private static final Map<String, MutableInt> CONNECTIONS = new HashMap<>();
+        private static final ReentrantLock lock = new ReentrantLock();
+        private static int totalConnectionNum = 0;
+
+        private final int maxConnections;
+        private final int maxConnectionPerIp;
+        private final boolean maxConnectionsLimitEnabled;
+        private final boolean maxConnectionsLimitPerIpEnabled;
+
+        public DefaultConnectionController(ServiceConfiguration configuration) {
+            this.maxConnections = configuration.getBrokerMaxConnections();
+            this.maxConnectionPerIp = configuration.getBrokerMaxConnectionsPerIp();
+            this.maxConnectionsLimitEnabled = configuration.getBrokerMaxConnections() > 0;
+            this.maxConnectionsLimitPerIpEnabled = configuration.getBrokerMaxConnectionsPerIp() > 0;
+        }
+
+        @Override
+        public boolean increaseConnection(SocketAddress remoteAddress) {
+            if (!maxConnectionsLimitEnabled && !maxConnectionsLimitPerIpEnabled) {
+                return true;
+            }
+            if (!(remoteAddress instanceof InetSocketAddress)
+                    || !isLegalIpAddress(((InetSocketAddress) remoteAddress).getHostString())) {
+                return true;
+            }
+            lock.lock();
+            try {
+                String ip = ((InetSocketAddress) remoteAddress).getHostString();
+                if (maxConnectionsLimitPerIpEnabled) {
+                    CONNECTIONS.computeIfAbsent(ip, (x) -> new MutableInt(0)).increment();
```

Comment:
If using `CONNECTIONS.computeIfAbsent` the connection count per IP will doesn't be increased after it is initialized.

Reply:
```
        Map<String, AtomicInteger> map = new HashMap<>();
        map.computeIfAbsent("1", (x) -> new AtomicInteger(0)).getAndIncrement();
        map.computeIfAbsent("1", (x) -> new AtomicInteger(0)).getAndIncrement();
        map.computeIfAbsent("1", (x) -> new AtomicInteger(0)).getAndIncrement();
        map.computeIfAbsent("1", (x) -> new AtomicInteger(0)).getAndIncrement();
        map.computeIfAbsent("1", (x) -> new AtomicInteger(0)).getAndIncrement();
        System.out.println(map.get("1").get());
```
the result is:
![image](https://user-images.githubusercontent.com/9758905/124586807-7d5c0300-de89-11eb-807d-f47fef50b8b1.png)

Reply:
Oh, sorry, I'm wrong, I thought the code `.getAndIncrement()` is in the `computeIfAbsent` function.